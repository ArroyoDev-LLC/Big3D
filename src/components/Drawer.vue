<script lang="ts">
import { defineComponent, ref } from 'vue'
export default defineComponent({
  name: 'Drawer',
  props: {
    title: {
      type: String,
      default: 'Drawer'
    },
    defaultState: {
      type: Boolean,
      default: false
    }
  },
  setup(props) {
    const toggle = ref(props.defaultState)
    return {
      toggle
    }
  }
})
</script>

<template>
  <slot name="header" :value="toggle">
    <div
      class="flex justify-between cursor-pointer border-dotted border-b border-opacity-40 py-1"
      @click="toggle = !toggle"
    >
      <slot name="title">
        <div class="text-white text-lg font-bold">
          {{ title }}
        </div>
      </slot>
      <slot name="toggle" :value="toggle">
        <button class="text-yellow text-xl font-bold">
          {{ toggle ? 'x' : '+' }}
        </button>
      </slot>
    </div>
  </slot>
  <div class="drawer" :class="toggle ? 'h-64 sm:h-48' : 'h-1'">
    <slot name="content" />
  </div>
</template>

<style scoped>
.drawer {
  @apply overflow-hidden;
  transition: all 0.2s;
}
</style>
