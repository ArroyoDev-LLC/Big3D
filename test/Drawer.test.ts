import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import Drawer from '../src/components/Drawer.vue'

describe('Drawer', () => {
  it('should render', () => {
    const wrapper = mount(Drawer)
    expect(wrapper.text()).toContain('Drawer')
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should toggle', async () => {
    const wrapper = mount(Drawer)
    expect(wrapper.find('.drawer').text()).toBe('')
    await wrapper.find('button').trigger('click')
    expect(wrapper.find('.drawer').text()).toBe('Content')
  })
  it('should set default value', () => {
    const wrapper = mount(Drawer, { props: { defaultState: true } })
    expect(wrapper.find('.drawer').text()).toBe('Content')
  })
  it('should set the title', () => {
    const wrapper = mount(Drawer, { props: { title: 'Title' } })
    expect(wrapper.find('.title').text()).toBe('Title')
  })
})
