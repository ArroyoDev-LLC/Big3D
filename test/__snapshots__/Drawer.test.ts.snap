// Vitest Snapshot v1

exports[`Drawer > should render 1`] = `
"<div class=\\"flex justify-between cursor-pointer border-dotted border-b border-opacity-40 py-1\\" data-v-a874e27e=\\"\\">
  <div class=\\"title\\" data-v-a874e27e=\\"\\">Drawer</div><button class=\\"text-yellow text-xl font-bold\\" data-v-a874e27e=\\"\\">+</button>
</div>
<div class=\\"drawer h-1\\" data-v-a874e27e=\\"\\">
  <!--v-if-->
</div>"
`;
